### PR TITLE
HP3478A - Add calibration data related functions.

### PR DESCRIPTION
**Pull request by tomverbeure**
The HP3478A stores its calibration data in SRAM, using a Lithium battery to retain the contents. To avoid calibration data loss, it's advised to back up this data.

This PR adds the following functionality:

* read the calibration data into an array
* write calibration data back to the device
* check the integrity of the calibration data

### CODE REVIEW

File: pymeasure/instruments/hp/hp3478A.py [hunk 1]
```diff
@@ -463,3 +463,88 @@ def trigger(self):
...
+
+        The CAL ENABLE switch at the front of the instrument must be set to ON,
+        otherwise an exception is raised.
+
+        IMPORTANT: changing the calibration data results in permanent loss of
+        the previous data. Use with care!
+
+        """
+
+        if not(self.calibration_enabled):
```
Comment (bmoneke) [inline pymeasure/instruments/hp/hp3478A.py:504]: I learnt it to write `not` as a keyword and not as a function:
```suggestion
        if not self.calibration_enabled:
```

File: pymeasure/instruments/hp/hp3478A.py [hunk 2]
```diff
@@ -463,3 +463,88 @@ def trigger(self):
...
+
+        Expects an array of 256 values between 0 and 15.
+
+        The CAL ENABLE switch at the front of the instrument must be set to ON,
+        otherwise an exception is raised.
+
+        IMPORTANT: changing the calibration data results in permanent loss of
+        the previous data. Use with care!
+
+        """
```
Comment (bmoneke) [inline pymeasure/instruments/hp/hp3478A.py:502]: This docstring works well as a method docstring. For a property, however, only the getter docstring is shown in the documentation such that you have to add this text to the getter docstring (obviously combining them in a meaningful way).
In the end, you want to have a single docstring describing the property as getter **and** setter.

File: pymeasure/instruments/hp/hp3478A.py [hunk 3]
```diff
@@ -463,3 +463,88 @@ def trigger(self):
...
+        cal_data = []
+        for addr in range(0, 256):
+            # To fetch one nibble: 'W<address>', where address is a raw 8-bit number.
+            cmd = bytes([ord('W'), addr])
+            self.write_bytes(cmd)
+            rvalue = self.read_bytes(1)[0]
+            if rvalue < 64 or rvalue >= 80:
+                raise Exception("calibration nibble out of range")
+            cal_data.append(rvalue-64)
```
Comment (bmoneke) [inline pymeasure/instruments/hp/hp3478A.py:486]: I first got confused, why do you check the range (64, 80) if values are in the range (0,15). A comment could help.
```suggestion
            # rvalue is a hexadecimal ASCII character.
            if rvalue < 64 or rvalue >= 80:
                raise Exception("calibration nibble out of range")
            cal_data.append(rvalue-64)
```

File: pymeasure/instruments/hp/hp3478A.py [hunk 4]
```diff
@@ -463,3 +463,88 @@ def trigger(self):
...
+
+        Returns True when the checksum of all used calibration entries is
+        correct.
+
+        """
+        for entry_nr in range(0, 19):
+            if entry_nr in [5, 16, 18]:
+                continue
+            if not(self.verify_calibration_entry(cal_data, entry_nr)):
+                return False
```
Comment (bmoneke) [inline pymeasure/instruments/hp/hp3478A.py:549]: I'm unsure, whether raising an exception would be better than just returning false. I guess it depends on the use case.

**Review by bmoneke — CHANGES_REQUESTED** [review]
Thanks for your contribution. You might add yourself to `AUTHORS.txt`.

I kindly ask you, to add unit tests for the verification and protocol tests for the data transfer, see [documentation](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html#protocol-tests). That ensures that it works as expected. If you have a device at hand, you might use #752 to get a copy of calibration data for the tests.

File: pymeasure/instruments/hp/hp3478A.py (hunk 4 above)
Comment (tomverbeure) [reply pymeasure/instruments/hp/hp3478A.py:549]: An exception would be IMO fine when you add this function as part of a non-explicit sanity check (like some of the other exception that I throw in the write data setter). 

But when used stand-alone the user wants to know the status, so a boolean makes more sense IMO?

Comment (bmoneke) [reply pymeasure/instruments/hp/hp3478A.py:549]: That's the reason I was unsure. As you describe it, a boolean might be better.

**Comment by LongnoseRob** [comment]
> The HP3478A stores its calibration data in SRAM, using a Lithium battery to retain the contents. To avoid calibration data loss, it's advised to back up this data.
> 
> This PR adds the following functionality:
> 
>     * read the calibration data into an array
> 
>     * write calibration data back to the device
> 
>     * check the integrity of the calibration data

Thanks for adding the calibration data handling methods, I have a HP3478A avialable for tests and verifications if required.
Will probably test it over the next weekend.

**Comment by tomverbeure** [comment]
> Thanks for adding the calibration data handling methods, I have a HP3478A avialable for tests and verifications if required. Will probably test it over the next weekend.

I did quite a bit of testing with [this test script](https://github.com/tomverbeure/hp3478a_calibration/blob/pymeasure/pymeasure_test.py), as part of writing [a blog post](https://tomverbeure.github.io/2022/12/02/HP3478A-Multimeter-Calibration-Data-Backup-and-Battery-Replacement.html)  (which will be updated to mention `pymeasure` once the PR has been merged.)

But you're more than welcome to give it a try yourself!

File: pymeasure/instruments/hp/hp3478A.py [hunk 5]
```diff
@@ -463,3 +463,88 @@ def trigger(self):
...
+        IMPORTANT: changing the calibration data results in permanent loss of
+        the previous data. Use with care!
+
+        """
+
+        if not(self.calibration_enabled):
+            raise Exception("CAL ENABLE switch not set to ON")
+        if len(cal_data) != 256:
+            raise Exception("cal_data must contain 256 values")
+
```
Comment (bilderbuchi) [inline pymeasure/instruments/hp/hp3478A.py:508]: Should you use the checksum verification from below on `cal_data` before starting the write?

**Comment by bilderbuchi** [comment]
By the way, if you don't want to be annoyed by the "pesky" linter, have your editor use the `flake8` (or `darker`, if you must) configuration we have in the repo, or if that's not possible, just run `flake8 .` before committing. :-)

File: pymeasure/instruments/hp/hp3478A.py (hunk 5 above)
Comment (bmoneke) [reply pymeasure/instruments/hp/hp3478A.py:508]: I also thought about that.
Its another check preventing erroneous change and thus loss of data.

Comment (tomverbeure) [reply pymeasure/instruments/hp/hp3478A.py:508]: I'll make the check enabled by default, with a function parameter to override: I definitely want to be able to write corrupt data to check that the write function is really working.

Comment (bmoneke) [reply pymeasure/instruments/hp/hp3478A.py:508]: Can't you test the write function with a working calibration data set? Maybe using two different sets to verify, that it changes?

**Comment by tomverbeure** [comment]
> By the way, if you don't want to be annoyed by the "pesky" linter, have your editor use the `flake8` (or `darker`, if you must) configuration we have in the repo, or if that's not possible, just run `flake8 .` before committing. :-)

I'm not annoyed about linting process, but about some linting rules.

For example, the fact that it complains about the first `True`:

```python
        assert instr.calibration_enabled == True
        assert instr.calibration_enabled == False
```

Anyway, it's a distraction not worth going further into.

**Comment by LongnoseRob** [comment]
@tomverbeure just gave this code a test-run with my HP3478A, reading and writing the cal-data works fine.
Thanks for the addition, now I can start looking into replacing the battery..

**Comment by tomverbeure** [comment]
> 

> @tomverbeure just gave this code a test-run with my HP3478A, reading and writing the cal-data works fine. Thanks for the addition, now I can start looking into replacing the battery..

Good to hear!

@bmoneke Is there anything else you'd like to see changed? I don't have a way to check the generated documentation, so maybe there could be some issue left there.

Tom

File: pymeasure/instruments/hp/hp3478A.py [hunk 6]
```diff
@@ -463,3 +463,109 @@ def trigger(self):
...
+        The cal_data parameter format is the same as the 'calibration_data' property.
+
+        Verification of the cal_data array can be bypassed by setting
+        'verify_calibration_data' to False.
+
+        """
+        if verify_calibration_data:
+            if not self.verify_calibration_data(cal_data):
+                raise Exception("cal_data verification fail.")
```
Comment (bmoneke) [inline pymeasure/instruments/hp/hp3478A.py:527]: You can (but don't have to) combine them:
```suggestion
        if verify_calibration_data and not self.verify_calibration_data(cal_data):
            raise Exception("cal_data verification fail.")
```
Only if the first part (`verify_calibration_data`) is True, the second call is done.

File: tests/instruments/hp/test_hp3478a.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        1, 12, 2, 1, 15, 10, 10, 0, 0, 0, 0, 4, 2, 3, 0, 0,
+        3, 12, 14, 7, 0, 0, 0, 0, 0, 4, 3, 0, 1, 12, 3, 14,
+        8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 15, 9, 9,
+        8, 6, 0, 9, 3, 14, 3, 1, 1, 12, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 15, 15, 0, 0, 0, 0, 0, 0, 0, 0
+    ]
+
+
+def convert_cal_data_to_cal_read_xfers(cal_data):
+    # Convert cal_data into 256 single byte 'W' read operations.
```
Comment (bmoneke) [inline tests/instruments/hp/test_hp3478a.py:49]: What are "xfers"? Transfer?

File: tests/instruments/hp/test_hp3478a.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                (b"B", b'\x00\x20\x00\x00\x00'),        # cal_enable bit is set
+                (b"B", b'\x00\x00\x00\x00\x00')
+            ],
+    ) as instr:
+        assert instr.calibration_enabled
+        assert not instr.calibration_enabled
+
+
+def test_calibration_data_getter():
+    cal_read_xfers = convert_cal_data_to_cal_read_xfers(VALID_CAL_DATA)
```
Comment (bmoneke) [inline tests/instruments/hp/test_hp3478a.py:83]: You use this line several times, why don't you make it a global variable (it is static) or a fixture?

File: tests/instruments/hp/test_hp3478a.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+            # setter pass
+            [
+                (b"B", b'\x00\x20\x00\x00\x00'),
+            ] + valid_cal_write_xfers +
+
+            # setter fail due to invalid data
+            [
+                (b"B", b'\x00\x20\x00\x00\x00'),
+            ]
```
Comment (bmoneke) [inline tests/instruments/hp/test_hp3478a.py:112]: I'd split it in three different tests, that way you know why the setter failed and not only, that it failed. If the first part fails, you do not know, whether the other parts work as expected.
In fact, pytest recommends to do a single assert in a test.

If you want to test several outcomes of some action, you can do the action in a fixture and have several tests, each testing one outcome. You can use classes (starting with "Test") to group tests together.

File: tests/instruments/hp/test_hp3478a.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+    ) as instr:
+        # There will be no data writes when cal_enable is false.
+        try:
+            instr.calibration_data = VALID_CAL_DATA
+        except Exception:
+            pass
```
Comment (bmoneke) [inline tests/instruments/hp/test_hp3478a.py:119]: Do you expect an exception to be raised? Then you can use `pytest.raises`:
```python
        with pytest.raises(SomeFunnyException, match="error message which I want to verify"):
            instr.calibration_data = VALID_CAL_DATA
```
That way the test passes if and only if `SomeFunnyException` was raised with a message containing above string. If the test fails, you know that raising the error did not happen as expected.

File: tests/instruments/hp/test_hp3478a.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+        # Writing correct data
+        instr.calibration_data = VALID_CAL_DATA
+
+        # Assigning invalid data results in an Exception without data writes
+        try:
+            valid_cal_data_corrupt = VALID_CAL_DATA.copy()
+            valid_cal_data_corrupt[1] = 1
```
Comment (bmoneke) [inline tests/instruments/hp/test_hp3478a.py:127]: You use it below as well. Why don't you make it as a global variable?

**Review by bmoneke — CHANGES_REQUESTED** [review]
I commented some tests. Thanks for including them. If you have questions regarding pytest, don't hesitate to ask.
For the one suggestion you might have to import pytest.

File: pymeasure/instruments/hp/hp3478A.py (hunk 6 above)
Comment (tomverbeure) [reply pymeasure/instruments/hp/hp3478A.py]: Fixed.

File: tests/instruments/hp/test_hp3478a.py (hunk 7 above)
Comment (tomverbeure) [reply tests/instruments/hp/test_hp3478a.py:51]: Yes.

File: tests/instruments/hp/test_hp3478a.py (hunk 8 above)
Comment (tomverbeure) [reply tests/instruments/hp/test_hp3478a.py:85]: VALID_CAL_DATA is the root calibration data that belongs in the instrument SRAM.

The 2 convert_* functions are functions that transform it into GPIB transfers. I think it makes sense to have those factored as individual functions. It's something that's useful to have if others want to reuse it at some point for their own code.

After that:
*  test_calibration_data_getter creates a cal_read_xfers array, 
*  test_calibration_data_setter_pass creates a valid_cal_write_xfers array from VALID_CAL_DATA
*  test_write_calibration_data creates invalid_cal_write_xfers from a corrupted VALID_CAL_DATA

So cases of anything being used multiple times.

File: tests/instruments/hp/test_hp3478a.py (hunk 9 above)
Comment (tomverbeure) [reply tests/instruments/hp/test_hp3478a.py]: Done.

File: tests/instruments/hp/test_hp3478a.py (hunk 10 above)
Comment (tomverbeure) [reply tests/instruments/hp/test_hp3478a.py]: Done.

File: tests/instruments/hp/test_hp3478a.py (hunk 11 above)
Comment (tomverbeure) [reply tests/instruments/hp/test_hp3478a.py:133]: See above. Only used once.

File: pymeasure/instruments/hp/hp3478A.py [hunk 12]
```diff
@@ -522,9 +522,8 @@ def write_calibration_data(self, cal_data, verify_calibration_data=True):
         'verify_calibration_data' to False.
 
         """
-        if verify_calibration_data:
-            if not self.verify_calibration_data(cal_data):
-                raise Exception("cal_data verification fail.")
+        if verify_calibration_data and not self.verify_calibration_data(cal_data):
+            raise Exception("cal_data verification fail.")
```
Comment (bmoneke) [inline pymeasure/instruments/hp/hp3478A.py:526]: As far as I know, you should not use 'Exception', but some subclass. Here I'd use a Value Error (use it in the tests as well).

Comment (tomverbeure) [reply pymeasure/instruments/hp/hp3478A.py]: Done.

**Comment by bmoneke** [comment]
@tomverbeure may I merge this PR?

**Comment by tomverbeure** [comment]
Of course! :-)

**Comment by bmoneke** [comment]
Thanks for your contribution!

**Pull request merged**